src/Lecture3.hs

```diff
@@ -103,85 +104,93 @@ have a lawful 'Monoid' instance.
 -}
 
 newtype Gold = Gold
-    { unGold :: Int
-    } deriving (Show, Eq)
+  { unGold :: Int
+  }
+  deriving (Show, Eq)
 
 -- | Addition of gold coins.
 instance Semigroup Gold where
-
+  (<>) :: Gold -> Gold -> Gold
+  (<>) (Gold x) (Gold y) = Gold (x + y)
 
 instance Monoid Gold where
+  mempty = Gold 0
 
-
-{- | A reward for completing a difficult quest says how much gold
-you'll receive and whether you'll get a special reward.
-
-If you combine multiple rewards, the final reward will contain a
-special prize if at least one of the rewards is special.
--}
+-- | A reward for completing a difficult quest says how much gold
+-- you'll receive and whether you'll get a special reward.
+--
+-- If you combine multiple rewards, the final reward will contain a
+-- special prize if at least one of the rewards is special.
 data Reward = Reward
-    { rewardGold    :: Gold
-    , rewardSpecial :: Bool
-    } deriving (Show, Eq)
+  { rewardGold :: Gold,
+    rewardSpecial :: Bool
+  }
+  deriving (Show, Eq)
 
 instance Semigroup Reward where
-
+  (<>) :: Reward -> Reward -> Reward
+  (<>) (Reward goldA specialA) (Reward goldB specialB) = Reward (goldA <> goldB) (specialA || specialB)
 
 instance Monoid Reward where
+  mempty = Reward (mempty :: Gold) False
```

Comment:
You don't need to have the type annotation here because Haskell has type inference and it works here
```suggestion
  mempty = Reward mempty False
```